File: go/0007.go
/* file: 0007.go
 * by  : agaric
 * copy: public domain
 * desc: project euler #7 - "10001st prime"
 * lang: go
 */

package main

import "fmt"

func sieve(n int) []int {
	a := []int{}
	b := make([]bool, n)
	for i, _ := range b {
		b[i] = true
	}
	b[0] = false
	if n >= 1 { b[1] = false }
	for i, x := range b {
		if x {
			m := i * i;
			for m < n {
				b[m] = false
				m += i;
			}
		}
	}
	for i, x := range b {
		if x { a = append(a, i) }
	}
	return a
}

func main() {
	fmt.Println(sieve(110000)[10000])
}

File: haskell/0007.hs
-- file: 0007.hs
-- by  : agaric
-- desc: project euler #7 - "10001st prime"

module Main (main) where

main :: IO ()
main = print answer

answer :: Integer
answer = primes !! 10000

primes :: [Integer]
primes = 2 : filter (null . tail . factors) [3, 5..]

factors :: Integer -> [Integer]
factors n = factor n primes

factor :: Integer -> [Integer] -> [Integer]
factor _ [] = []
factor n (p:ps)
  | p * p > n    = [n]
  | rem n p == 0 = p : factor (div n p) (p:ps)
  | otherwise    = factor n ps
